jclouds-plugin/src/main/java/jenkins/plugins/jclouds/compute/JCloudsCloud.java

```diff
@@ -100,17 +121,48 @@ public static JCloudsCloud getByName(String name) {
         return (JCloudsCloud) Hudson.getInstance().clouds.getByName(name);
     }
 
+    public String getCloudGlobalKeyId() {
+        return cloudGlobalKeyId;
+    }
+
+    public void setCloudGlobalKeyId(final String value) {
+        cloudGlobalKeyId = value;
+    }
+
+    public String getGlobalPrivateKey() {
+        if (Strings.isNullOrEmpty(cloudGlobalKeyId)) {
+            return "";
+        }
+        SSHUserPrivateKey supk = CredentialsMatchers.firstOrNull(
+                CredentialsProvider.lookupCredentials(SSHUserPrivateKey.class, Hudson.getInstance(), ACL.SYSTEM, null),
+                CredentialsMatchers.withId(cloudGlobalKeyId));
+        if (null != supk) {
+            return supk.getPrivateKey();
+        }
+        return "";
+    }
+
+    public String getGlobalPublicKey() {
+        try {
+            return SSHPublicKeyExtractor.extract(getGlobalPrivateKey(), null);
+        } catch (IOException e) {
+            LOGGER.warning(String.format("Error while extracting public key: %s", e));
+        }
+        return "";
+    }
+
     @DataBoundConstructor
-    public JCloudsCloud(final String profile, final String providerName, final String identity, final String credential, final String privateKey,
-                        final String publicKey, final String endPointUrl, final int instanceCap, final int retentionTime, final int scriptTimeout, final int startTimeout,
-                        final String zones, final List<JCloudsSlaveTemplate> templates) {
+    public JCloudsCloud(final String profile, final String providerName, final String identity, final String credential, final String cloudGlobalKeyId,
+            final String endPointUrl, final int instanceCap, final int retentionTime, final int scriptTimeout, final int startTimeout,
+            final String zones, final List<JCloudsSlaveTemplate> templates) {
         super(Util.fixEmptyAndTrim(profile));
         this.profile = Util.fixEmptyAndTrim(profile);
         this.providerName = Util.fixEmptyAndTrim(providerName);
         this.identity = Util.fixEmptyAndTrim(identity);
         this.credential = Secret.fromString(credential);
-        this.privateKey = privateKey;
-        this.publicKey = publicKey;
+        this.privateKey = null; // No longer used
```

Comment:
Instead of @Deprecated, I'm going to make them private instead. The persistence layer does not care about access restrictions anyway.
Regarding volatile: Why?